Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -826,6 +861,66 @@ Timestamp Serializer::DecodeTimestamp(
   return Timestamp{timestamp_proto.seconds, timestamp_proto.nanos};
 }
 
+/* static */
+google_type_LatLng Serializer::EncodeGeoPoint(const GeoPoint& geo_point_value) {
+  google_type_LatLng result{};
+  result.latitude = geo_point_value.latitude();
+  result.longitude = geo_point_value.longitude();
+  return result;
+}
+
+/* static */
+GeoPoint Serializer::DecodeGeoPoint(nanopb::Reader* reader,
+                                    const google_type_LatLng& latlng_proto) {
+  // The GeoPoint ctor will assert if we provide values outside the valid range.
+  // However, since we're decoding, a single corrupt byte could cause this to
+  // occur, so we'll verify the ranges before passing them in since we'd rather
+  // not abort in these situations.
+  double latitude = latlng_proto.latitude;
+  double longitude = latlng_proto.longitude;
+  if (std::isnan(latitude) || latitude < -90 || 90 < latitude) {
+    reader->Fail("Invalid message: Latitude must be in the range of [-90, 90]");
+  } else if (std::isnan(longitude) || longitude < -180 || 180 < longitude) {
+    reader->Fail(
+        "Invalid message: Latitude must be in the range of [-180, 180]");
+  }
+
+  if (!reader->status().ok()) return GeoPoint();
+  return GeoPoint(latitude, longitude);
+}
+
+/* static */
+google_firestore_v1_ArrayValue Serializer::EncodeArray(
+    const std::vector<FieldValue>& array_value) {
+  google_firestore_v1_ArrayValue result{};
+
+  size_t count = array_value.size();
+  HARD_ASSERT(count <= std::numeric_limits<pb_size_t>::max(),
```

Comment:
Very optional: looks like nanopb defines a `PB_SIZE_MAX` constant for this?

(note that while this check is valid for nanopb containers, it's not necessarily valid for containers in general, which is why it might be a little better to do the check using some container-specific constants. For a standard library container, the maximum size is bound by its `difference_type` (otherwise, it would be possible to create containers with elements that are not addressable), which is signed and consequently might be much lower than the maximum `size_type`. This is pretty pedantic, though -- feel free to ignore).

Reply:
> PB_SIZE_MAX

Done (here and throughout this file; one other spot.)

> container size/pedantic/etc

Honestly, I was more worried about the unsigned -> signed conversion and dodging the compiler error/warning that resulted from it. But good point about container size. (Realistically, if someone requests a container that exceeds the difference_type bound, there's unlikely to be enough ram to accommodate that. But in cases where we need to do some sort of comparison anyways, we might as well use the correct one.)